### PR TITLE
Added a configuration option for removing newlines as you type

### LINKED ISSUE #93: Add config option for Clear Blank Line on New Line
Specific Function: https://github.com/atom/whitespace/blob/master/lib/whitespace.coffee#L34

Currently, when the option to strip whitespace from whitespace only lines on save is enabled, the editor does this as you're typing, which I find to be very unhelpful when I'm attempting to edit code I've just written.

Could we please have a configuration option for this change?

### PR DESCRIPTION
**Pull request by Lixquid**
Fixes #93.

Adds a configuration option for the intrusive "Strip Newlines as you type" functionality.

### CODE REVIEW

File: lib/main.coffee [hunk 1]
```diff
@@ -8,6 +8,9 @@ module.exports =
       scopes:
         '.source.jade':
           default: false
+    removeTrailingWhitespaceAsYouType:
```
Comment (50Wliu) [inline lib/main.coffee]: Maybe `activelyRemoveTrailingWhitespace`?

Comment (Lixquid) [inline lib/main.coffee]: I'd be open to any name, although the association to `removeTrailingWhitespace` is useful, as it doesn't strip whitespace if that isn't enabled, regardless of this setting.

**Comment by Lixquid** [comment]
Could I please have at least some visibility / response on this? Up until now, I've had a workaround to stop this from happening, but it's recently stopped working, and this is really, really intrusive to my workflow. (@50Wliu)

File: lib/main.coffee [hunk 2]
```diff
@@ -9,6 +9,10 @@ module.exports =
         '.source.jade':
           default: false
       description: 'Automatically remove whitespace characters at ends of lines when the buffer is saved. To disable/enable for a certain language, use [syntax-scoped properties](https://github.com/atom/whitespace#readme) in your `config.cson`.'
+    removeTrailingWhitespaceAsYouType:
+      type: 'boolean'
+      default: false
```
Comment (lee-dohm) [inline lib/main.coffee]: The current behavior is `true` correct? So this should default to `true`. We don't want to break current default behavior.

Comment (Lixquid) [inline lib/main.coffee]: While I disagree that the setting should be enabled by default, I agree that it's more important to preserve previous defaults. I've committed the change.

**Comment by maxbrunsfeld** [comment]
@Lixquid Sorry that this PR was ignored for so long. Thanks for speaking up about getting a review. A few high-level questions:
* Out of curiosity, could you describe in a little more detail why you find the current behavior to be really intrusive to your workflow? I'm wondering if there's any room to improve the feature, rather than adding an option to turn it off.
* What should happen when `removeTrailingWhitespace` is `false`, but `removeTrailingWhitespaceAsYouType` is `true`? It looks like currently, it won't remove trailing whitespace as the user types. I'm concerned this might be a little confusing for people. Maybe we should rename the first option to `removeTrailingWhitespaceOnSave`, and make the two behaviors completely independent.

**Comment by lee-dohm** [comment]
I'm concerned with maintaining the current functionality for the users who have come to expect it. Whatever solution we come up with, the behavior shouldn't change out from under people. If configuration options change meaning or naming ... we need to build conversion utilities so that the migration is seamless.

**Comment by Lixquid** [comment]
Of course, I understand big projects like Atom can be very hard to manage.

In regards to my personal workflow, I tend to input a lot of whitespace into my code, both vertically and horizontally. My eyesight isn't too great, so I tend to "lose" what I'm working on if I don't surround important logical blocks with blank lines. Because of this, I normally start blocks by creating a "gap" in the code for me to work in, indented to the correct amount, which gets quickly destroyed by automatic whitespace removal. It's likely niche and local to only a few people, but it's a fairly large leap to go from on the fly removal to prettifying / formatting on save.

In regards to your comment about renaming `removeTrailingWhitespace` to `removeTrailingWhitespaceOnSave`, I agree that it would make settings significantly easier to understand, although it's possible it might be confusing for people that have already changed `removeTrailingWhitespace` from its default value.

**Comment by lee-dohm** [comment]
Perhaps the right thing here is to add this functionality and then do a separate PR for the renaming and migration of configuration settings?

**Comment by Lixquid** [comment]
I think that's a good idea, we can separate discussion about the viability of the feature versus how it should be presented.

**Comment by benogle** [comment]
IMO, this adds complexity to the already probably-confusing options. Here's what I think we should do:

Change `removeTrailingWhitespace` to be an enum config type that accepts three options: `'on-save', 'on-press-enter', and 'never'`. 

That would resolve the clarity issue, and keep this contained to one option. Additionally, the default should continue to be `on-save`. People already freak out about this default, and being more forceful when typing would not have a warm reception.

**Comment by maxbrunsfeld** [comment]
> Additionally, the default should continue to be `on-save`

Actually, the default is currently to remove trailing whitespace when pressing enter. I don't think there's been too much complaint about it.

**Comment by benogle** [comment]
Ok, so that only happens on whitespace only lines, not trailing whitespace on all lines. I'll admit, then I dont understand the premise of this PR. I thought this was stripping trailing space on non-empty lines when typing, which now happens on save.

**Comment by benogle** [comment]
Alright, alright, I get it. Disregard my last comment. Sorry for the noise. 

These options are confusing, and this behavior is nuanced. I talked to @maxbrunsfeld about this and we came up with a slightly different approach:

1. Remove the `removeTrailingWhitespace` option's effect on this behavior; `removeTrailingWhitespace` can be false, and this new option can be true to get the current behavior.
1. Keep this option, but rename it to something like `clearPreviousWhitespaceOnlyLineOnEnter` for clarity

**Comment by Lixquid** [comment]
Superseded.

**Pull request closed without merging**